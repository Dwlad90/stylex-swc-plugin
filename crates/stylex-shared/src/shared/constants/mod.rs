pub(crate) mod application_order;
pub(crate) mod common;
pub(crate) mod cursor_flip;
pub(crate) mod evaluation_errors;
pub(crate) mod legacy_expand_shorthands_order;
pub(crate) mod length_units;
pub(crate) mod long_hand_logical;
pub(crate) mod long_hand_physical;
pub mod messages;
pub(crate) mod number_properties;
pub(crate) mod priorities;
pub(crate) mod property_specificity_order;
pub(crate) mod shorthands_of_longhands;
pub(crate) mod shorthands_of_shorthands;
pub(crate) mod time_units;
pub(crate) mod unitless_number_properties;
