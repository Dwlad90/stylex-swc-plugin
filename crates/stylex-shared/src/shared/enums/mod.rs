pub(crate) mod core;
pub mod data_structures;
pub(crate) mod js;
pub(crate) mod misc;
