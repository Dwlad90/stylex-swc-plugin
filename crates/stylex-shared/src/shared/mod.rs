pub mod constants;
pub mod enums;
pub mod regex;
pub mod structures;
pub(crate) mod swc;
pub(crate) mod transformers;
pub mod utils;
