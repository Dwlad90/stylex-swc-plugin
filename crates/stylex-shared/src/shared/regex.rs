use once_cell::sync::Lazy;
use regex::Regex;

pub(crate) static INCLUDED_IDENT_REGEX: Lazy<Regex> =
  Lazy::new(|| Regex::new(r"__included_\d+__").unwrap());

pub(crate) static CSS_RULE_REGEX: Lazy<Regex> =
  Lazy::new(|| Regex::new(r"\w+:\s*([^;}]+);?").unwrap());

pub(crate) static WHITESPACE_NORMALIZER_MATH_SIGNS_REGEX: Lazy<Regex> =
  Lazy::new(|| Regex::new(r"(\d+%?|\))\s*([+\-*/%])\s*(\d*\.?\d+|\()").unwrap());

pub(crate) static SANITIZE_CLASS_NAME_REGEX: Lazy<Regex> =
  Lazy::new(|| Regex::new(r"[^.a-zA-Z0-9_-]").unwrap());

pub(crate) static WHITESPACE_BRACKET_NORMALIZER_REGEX: Lazy<Regex> =
  Lazy::new(|| Regex::new(r#"(\))(\S)|(\")(\")"#).unwrap());

pub(crate) static WHITESPACE_FUNC_NORMALIZER_REGEX: Lazy<Regex> =
  Lazy::new(|| Regex::new(r#"\(\s*([^)]*?)\s*\)\s*,\s*"#).unwrap());

pub(crate) static HASH_WHITESPACE_NORMALIZER_REGEX: Lazy<Regex> =
  Lazy::new(|| Regex::new(r"(\S)#").unwrap());

pub(crate) static LENGTH_UNIT_TESTER_REGEX: Lazy<Regex> =
  Lazy::new(|| Regex::new(r"^-?\d+(px|%|em|rem|ex|ch|vh|vw|vmin|vmax)?$").unwrap());

pub(crate) static CSS_URL_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r#"url\(.+\)"#).unwrap());

pub(crate) static CSS_PROPERTY_KEY: Lazy<Regex> =
  Lazy::new(|| Regex::new(r#"(\))(\S)|(\")(\")"#).unwrap());

pub(crate) static CLEAN_CSS_VAR: Lazy<Regex> = Lazy::new(|| Regex::new(r#"\\3(\d) "#).unwrap());

pub(crate) static IS_CSS_VAR: Lazy<Regex> =
  Lazy::new(|| Regex::new(r#"^var\(--[a-zA-Z0-9-_]+\)$"#).unwrap());

pub(crate) static MANY_SPACES: Lazy<Regex> = Lazy::new(|| Regex::new(r#"\s+"#).unwrap());

pub(crate) static WHITESPACE_NORMALIZER_EXTRA_SPACES_REGEX: Lazy<Regex> = Lazy::new(|| {
  Regex::new(
    r#"(?x)
      # Empty string quotes
      ^(\")\s(\")$ |
      # Multiple closing parens
      (\))\s+(\))
  "#,
  )
  .unwrap()
});

pub(crate) static DASHIFY_REGEX: Lazy<Regex> =
  Lazy::new(|| Regex::new(r"(^|[a-z])([A-Z])").unwrap());
