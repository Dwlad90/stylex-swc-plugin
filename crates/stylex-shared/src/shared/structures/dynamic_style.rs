use swc_core::ecma::ast::Expr;

#[derive(Debug, Clone)]
pub(crate) struct DynamicStyle {
  pub(crate) expression: Expr,
  pub(crate) key: String,
  pub(crate) path: String,
}
