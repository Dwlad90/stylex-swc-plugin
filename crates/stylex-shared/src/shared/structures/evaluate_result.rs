use std::collections::HashMap;

use indexmap::IndexMap;
use swc_core::{
  atoms::Atom,
  ecma::ast::{Expr, KeyValueProp, Lit},
};

use crate::shared::enums::data_structures::evaluate_result_value::EvaluateResultValue;

use super::{
  functions::FunctionConfig,
  theme_ref::ThemeRef,
  types::{DynamicFns, EvaluationCallback, TInlineStyles},
};

#[derive(Debug, Clone, PartialEq)]
pub struct EvaluateResult {
  pub(crate) confident: bool,
  pub value: Option<Box<EvaluateResultValue>>,
  pub(crate) deopt: Option<Box<Expr>>,
  pub(crate) inline_styles: Option<TInlineStyles>,
  pub(crate) fns: Option<DynamicFns>,
}

impl EvaluateResultValue {
  pub fn as_expr(&self) -> Option<&Expr> {
    match self {
      EvaluateResultValue::Expr(value) => Some(value),
      _ => None,
    }
  }

  pub fn as_vec(&self) -> Option<&Vec<Option<EvaluateResultValue>>> {
    match self {
      EvaluateResultValue::Vec(value) => Some(value),
      _ => None,
    }
  }

  pub fn as_map(&self) -> Option<&IndexMap<Box<Expr>, Vec<KeyValueProp>>> {
    match self {
      EvaluateResultValue::Map(value) => Some(value),
      _ => None,
    }
  }

  pub fn as_entries(&self) -> Option<&IndexMap<Box<Lit>, Box<Lit>>> {
    match self {
      EvaluateResultValue::Entries(value) => Some(value),
      _ => None,
    }
  }

  pub fn as_function(&self) -> Option<&FunctionConfig> {
    match self {
      EvaluateResultValue::FunctionConfig(value) => Some(value),
      _ => None,
    }
  }

  pub fn as_function_map(&self) -> Option<&HashMap<Atom, FunctionConfig>> {
    match self {
      EvaluateResultValue::FunctionConfigMap(value) => Some(value),
      _ => None,
    }
  }

  pub fn as_callback(&self) -> Option<&EvaluationCallback> {
    match self {
      EvaluateResultValue::Callback(value) => Some(value),
      _ => None,
    }
  }

  pub fn as_theme_ref(&self) -> Option<&ThemeRef> {
    match self {
      EvaluateResultValue::ThemeRef(value) => Some(value),
      _ => None,
    }
  }
}
