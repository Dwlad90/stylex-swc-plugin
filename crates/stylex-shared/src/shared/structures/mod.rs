pub(crate) mod application_order;
pub(crate) mod base_css_type;
pub(crate) mod dynamic_style;
pub mod evaluate_result;
pub mod functions;
pub(crate) mod injectable_style;
pub(crate) mod inline_style;
pub(crate) mod legacy_expand_shorthands_order;
pub(crate) mod member_transform;
pub mod meta_data;
pub mod named_import_source;
pub(crate) mod null_pre_rule;
pub(crate) mod order;
pub(crate) mod order_pair;
pub(crate) mod pair;
pub mod plugin_pass;
pub(crate) mod pre_rule;
pub(crate) mod pre_rule_set;
pub(crate) mod property_specificity;
pub(crate) mod property_specificity_order;
pub(crate) mod seen_value;
pub(crate) mod shorthands_of_shorthands;
pub(crate) mod state;
pub mod state_manager;
pub mod stylex_options;
pub(crate) mod stylex_state_options;
pub(crate) mod tests;
pub(crate) mod theme_ref;
pub(crate) mod types;
pub(crate) mod uid_generator;
