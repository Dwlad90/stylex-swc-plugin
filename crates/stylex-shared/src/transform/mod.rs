use rustc_hash::FxHashSet;
use swc_core::{
  common::{Mark, comments::Comments},
  ecma::{
    ast::{CallExpr, Callee, Expr, Id, MemberProp, Pass, VarDeclarator},
    transforms::base::resolver,
    utils::drop_span,
    visit::fold_pass,
  },
};

use crate::{
  StyleXOptionsParams,
  shared::{
    enums::core::TransformationCycle,
    structures::{
      named_import_source::{ImportSources, RuntimeInjection},
      plugin_pass::PluginPass,
      state_manager::StateManager,
      stylex_options::{CheckModuleResolution, StyleXOptions},
    },
    utils::common::increase_ident_count,
  },
};

mod fold;
pub(crate) mod styleq;
pub(crate) mod stylex;

pub struct StyleXTransform<C>
where
  C: Comments,
{
  pub comments: C,
  props_declaration: Option<Id>,
  pub state: StateManager,
}

impl<C> StyleXTransform<C>
where
  C: Comments,
{
  pub fn new(comments: C, plugin_pass: PluginPass, config: &mut StyleXOptionsParams) -> Self {
    let stylex_imports = fill_stylex_imports(&Some(config));

    let mut state = StateManager::new(config.clone().into());

    state.stylex_import.clone_from(&stylex_imports);

    state.options.import_sources = stylex_imports.into_iter().collect();

    state._state = plugin_pass;

    StyleXTransform {
      comments,
      props_declaration: None,
      state,
    }
  }

  pub fn new_test_force_runtime_injection(
    comments: C,
    plugin_pass: PluginPass,
    config: Option<&mut StyleXOptionsParams>,
  ) -> Self {
    let stylex_imports = fill_stylex_imports(&config);

    let mut state = match config {
      Some(config) => {
        config.runtime_injection = Some(true);
        config.treeshake_compensation = Some(true);

        StateManager::new(config.clone().into())
      }
      None => {
        let config = StyleXOptions {
          runtime_injection: RuntimeInjection::Boolean(true),
          treeshake_compensation: Some(true),
          unstable_module_resolution: CheckModuleResolution::Haste(
            StyleXOptions::get_haste_module_resolution(None),
          ),
          ..Default::default()
        };

        StateManager::new(config)
      }
    };

    state.options.import_sources = stylex_imports.into_iter().collect();

    state._state = plugin_pass;

    StyleXTransform {
      comments,
      props_declaration: None,
      state,
    }
  }

  pub fn new_test_force_runtime_injection_with_pass(
    comments: C,
    plugin_pass: PluginPass,
    config: Option<&mut StyleXOptionsParams>,
  ) -> impl Pass + use<C> {
    (
      resolve_factory(),
      fold_pass(Self::new_test_force_runtime_injection(
        comments,
        plugin_pass,
        config,
      )),
    )
  }

  pub fn new_test(
    comments: C,
    plugin_pass: PluginPass,
    config: Option<&mut StyleXOptionsParams>,
  ) -> Self {
    let stylex_imports = fill_stylex_imports(&config);

    let mut state = match config {
      Some(config) => StateManager::new(config.clone().into()),
      None => {
        let config = StyleXOptions {
          runtime_injection: RuntimeInjection::Boolean(false),
          treeshake_compensation: Some(true),
          class_name_prefix: "x".to_string(),
          unstable_module_resolution: CheckModuleResolution::Haste(
            StyleXOptions::get_haste_module_resolution(None),
          ),
          ..Default::default()
        };

        StateManager::new(config)
      }
    };

    state.options.import_sources = stylex_imports.into_iter().collect();

    state._state = plugin_pass;

    StyleXTransform {
      comments,
      props_declaration: None,
      state,
    }
  }
  pub fn new_test_with_pass(
    comments: C,
    plugin_pass: PluginPass,
    config: Option<&mut StyleXOptionsParams>,
  ) -> impl Pass + use<C> {
    (
      resolve_factory(),
      fold_pass(Self::new_test(comments, plugin_pass, config)),
    )
  }

  pub(crate) fn process_declaration(&mut self, call_expr: &mut CallExpr) -> Option<(Id, String)> {
    let state = &self.state;

    let stylex_imports = state.stylex_import_stringified();
    if let Callee::Expr(callee) = &mut call_expr.callee {
      match callee.as_ref() {
        Expr::Ident(ident) => {
          let ident_id = ident.to_id();

          if stylex_imports.contains(&ident.sym.to_string())
            || (state.cycle == TransformationCycle::TransformEnter
              && (state.stylex_create_import.contains(&ident.sym))
              || state.stylex_props_import.contains(&ident.sym)
              || state.stylex_keyframes_import.contains(&ident.sym)
              || state.stylex_first_that_works_import.contains(&ident.sym)
              || state.stylex_types_import.contains(&ident.sym)
              || state.stylex_create_theme_import.contains(&ident.sym)
              || state.stylex_define_vars_import.contains(&ident.sym)
              || state.stylex_attrs_import.contains(&ident.sym))
          {
            increase_ident_count(&mut self.state, ident);

            return Some((ident_id.clone(), format!("{}", ident.sym)));
          }
        }
        Expr::Member(member) => {
          if let Expr::Ident(ident) = member.obj.as_ref() {
            let ident_id = ident.to_id();

            if stylex_imports.contains(&ident.sym.to_string())
              || (state.cycle == TransformationCycle::TransformEnter
                && (state.stylex_create_import.contains(&ident.sym))
                || state.stylex_props_import.contains(&ident.sym)
                || state.stylex_keyframes_import.contains(&ident.sym)
                || state.stylex_first_that_works_import.contains(&ident.sym)
                || state.stylex_create_theme_import.contains(&ident.sym)
                || state.stylex_types_import.contains(&ident.sym)
                || state.stylex_define_vars_import.contains(&ident.sym)
                || state.stylex_attrs_import.contains(&ident.sym))
            {
              if let MemberProp::Ident(ident) = &member.prop {
                return Some((ident_id.clone(), format!("{}", ident.sym)));
              }
            }
          }
        }
        _ => {}
      }
    }

    None
  }

  pub(crate) fn transform_call_expression(&mut self, expr: &mut Expr) -> Option<Expr> {
    if let Expr::Call(ex) = expr {
      let declaration = self.process_declaration(ex);

      if declaration.is_some() {
        return self.transform_call_expression_to_stylex_expr(ex);
      }
    }

    None
  }

  pub(crate) fn get_call_var_name(
    &mut self,
    call: &CallExpr,
  ) -> (Option<String>, Option<VarDeclarator>) {
    let mut var_name: Option<String> = None;

    let parent_var_decl = self
      .state
      .declarations
      .iter()
      .find(|decl| match &decl.init {
        Some(init) => {
          if let Expr::Call(init_call) = init.as_ref() {
            init_call == &drop_span(call.clone())
          } else {
            false
          }
        }
        _ => false,
      })
      .cloned();

    if let Some(ref parent_var_decl) = parent_var_decl {
      if let Some(ident) = parent_var_decl.name.as_ident() {
        var_name = Some(ident.sym.to_string());
      }
    }

    (var_name, parent_var_decl)
  }
}

fn fill_stylex_imports(config: &Option<&mut StyleXOptionsParams>) -> FxHashSet<ImportSources> {
  let mut stylex_imports = FxHashSet::default();

  stylex_imports.insert(ImportSources::Regular("stylex".to_string()));
  stylex_imports.insert(ImportSources::Regular("@stylexjs/stylex".to_string()));

  if let Some(stylex_imports_extends) = match config {
    Some(config) => config.import_sources.clone(),
    None => None,
  } {
    stylex_imports.extend(stylex_imports_extends)
  }

  stylex_imports
}

#[warn(clippy::extra_unused_type_parameters)]
fn resolve_factory() -> impl Pass {
  let unresolved_mark = Mark::new();
  let top_level_mark = Mark::new();

  resolver(unresolved_mark, top_level_mark, true)
}
