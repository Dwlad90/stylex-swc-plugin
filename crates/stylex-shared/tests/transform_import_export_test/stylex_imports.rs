use crate::utils::transform::stringify_js;
use insta::assert_snapshot;
use stylex_shared::shared::structures::{
  named_import_source::{ImportSources, NamedImportSource},
  plugin_pass::PluginPass,
  stylex_options::{StyleXOptions, StyleXOptionsParams},
};
use swc_core::{
  common::FileName,
  ecma::parser::{Syntax, TsSyntax},
};

fn get_default_opts() -> StyleXOptionsParams {
  StyleXOptionsParams {
    unstable_module_resolution: Some(StyleXOptions::get_common_js_module_resolution(Some(
      "/stylex/packages/".to_string(),
    ))),
    ..StyleXOptionsParams::default()
  }
}

#[derive(Debug, Clone)]
struct FixtureImportMap {
  create: &'static str,
  create_theme: &'static str,
  define_consts: &'static str,
  define_vars: &'static str,
  first_that_works: &'static str,
  keyframes: &'static str,
  position_try: &'static str,
  props: &'static str,
}

impl Default for FixtureImportMap {
  fn default() -> Self {
    Self {
      create: "stylex.create",
      create_theme: "stylex.createTheme",
      define_consts: "stylex.defineConsts",
      define_vars: "stylex.defineVars",
      first_that_works: "stylex.firstThatWorks",
      keyframes: "stylex.keyframes",
      position_try: "stylex.positionTry",
      props: "stylex.props",
    }
  }
}

/// Rust version of the JS createStylesFixture
fn create_styles_fixture(
  import_text: Option<&str>,
  import_source: Option<&str>,
  import_map: Option<FixtureImportMap>,
) -> String {
  let import_text = import_text.unwrap_or("* as stylex");
  let import_source = import_source.unwrap_or("@stylexjs/stylex");
  let import_map = import_map.unwrap_or_default();

  let from = import_source;

  // Generate the defineConsts and defineVars output first (like JS version)
  let define_consts_and_vars_input = format!(
    r#"import {} from '{}';
export const constants = {}({{
  mediaQuery: '@media (min-width: 768px)',
}});
export const vars = {}({{
  bar: 'left'
}});
"#,
    import_text, from, import_map.define_consts, import_map.define_vars
  );

  let define_consts_and_vars_output = transform_with_options(
    &define_consts_and_vars_input,
    Some(&mut StyleXOptionsParams {
      unstable_module_resolution: Some(StyleXOptions::get_common_js_module_resolution(Some(
        "/stylex/packages/".to_string(),
      ))),
      ..StyleXOptionsParams::default()
    }),
    Some(FileName::Real("/stylex/packages/vars.stylex.js".into())),
    None,
  );

  let formatted = format!(
    r#"{define_consts_and_vars_output}
        const fallback1 = {position_try}({{
          anchorName: '--myAnchor',
          positionArea: 'top left',
        }});
        const fallback2 = {position_try}({{
          anchorName: '--otherAnchor',
          top: 'anchor(bottom)',
          insetInlineStart: 'anchor(start)',
        }});
        const styles = {create}({{
          root: {{
            animationName: {keyframes}({{
              from: {{
                backgroundColor: 'yellow'
              }},
              to: {{
                backgroundColor: 'orange'
              }},
            }}),
            positionTryFallbacks: `${{fallback1}}, ${{fallback2}}`,
            color: {{
              default: 'red',
              [constants.mediaQuery]: 'blue'
            }},
            position: {first_that_works}('sticky', 'fixed')
          }}
        }});

        const theme = {create_theme}(vars, {{
          bar: 'green'
        }});

        {props}(styles.root, theme);
    "#,
    define_consts_and_vars_output = define_consts_and_vars_output,
    position_try = import_map.position_try,
    create = import_map.create,
    keyframes = import_map.keyframes,
    first_that_works = import_map.first_that_works,
    create_theme = import_map.create_theme,
    props = import_map.props,
  );

  formatted
}

fn transform_with_options(
  input: &str,
  options: Option<&mut StyleXOptionsParams>,
  filename: Option<FileName>,
  import_sources: Option<Vec<ImportSources>>,
) -> String {
  let mut opts = options.cloned().unwrap_or_else(get_default_opts);
  if let Some(sources) = import_sources {
    opts.import_sources = Some(sources);
  }

  let syntax = Syntax::Typescript(TsSyntax {
    tsx: true,
    ..Default::default()
  });
  stringify_js(input, syntax, |tr| {
    stylex_shared::StyleXTransform::new_test_with_pass(
      tr.comments.clone(),
      PluginPass {
        cwd: None,
        filename: filename.unwrap_or(FileName::Real("/stylex/packages/TestFile.js".into())),
      },
      Some(&mut opts),
    )
  })
}

fn transform_and_snapshot(input: &str, snapshot_name: &str) {
  let opts = get_default_opts();
  let syntax = Syntax::Typescript(TsSyntax {
    tsx: true,
    ..Default::default()
  });
  let output = stringify_js(input, syntax, |tr| {
    stylex_shared::StyleXTransform::new_test_with_pass(
      tr.comments.clone(),
      stylex_shared::shared::structures::plugin_pass::PluginPass::default(),
      Some(&mut opts.clone()),
    )
  });
  assert_snapshot!(snapshot_name, output);
}

// --- import: none ---
#[test]
fn import_none() {
  let input = r#"
        export const styles = stylex.create({
          root: {
            color: 'red'
          }
        });
    "#;
  transform_and_snapshot(input, "import_none");
}

// --- import: non-stylex ---
#[test]
fn import_non_stylex() {
  let input = r#"
        import {foo, bar} from 'other';
    "#;
  transform_and_snapshot(input, "import_non_stylex");
}

// --- require: non-stylex ---
#[test]
fn require_non_stylex() {
  let input = r#"
        const {foo, bar} = require('other');
    "#;
  transform_and_snapshot(input, "require_non_stylex");
}

// --- import: wildcard (the default) ---
#[test]
fn import_wildcard_the_default() {
  let fixture = create_styles_fixture(None, None, None);
  transform_and_snapshot(&fixture, "import_wildcard_the_default");
}

// --- import: wildcard (non-stylex name) ---
#[test]
fn import_wildcard_non_stylex_name() {
  let fixture = create_styles_fixture(
    Some("* as foo"),
    None,
    Some(FixtureImportMap {
      create: "foo.create",
      create_theme: "foo.createTheme",
      define_consts: "foo.defineConsts",
      define_vars: "foo.defineVars",
      first_that_works: "foo.firstThatWorks",
      keyframes: "foo.keyframes",
      position_try: "foo.positionTry",
      props: "foo.props",
    }),
  );
  transform_and_snapshot(&fixture, "import_wildcard_non_stylex_name");
}

// --- import: named ---
#[test]
fn import_named() {
  let fixture = create_styles_fixture(
    Some(
      "{create, createTheme, defineConsts, defineVars, firstThatWorks, keyframes, positionTry, props}",
    ),
    None,
    Some(FixtureImportMap {
      create: "create",
      create_theme: "createTheme",
      define_consts: "defineConsts",
      define_vars: "defineVars",
      first_that_works: "firstThatWorks",
      keyframes: "keyframes",
      position_try: "positionTry",
      props: "props",
    }),
  );
  transform_and_snapshot(&fixture, "import_named");
}

// --- import: named alias ---
#[test]
fn import_named_alias() {
  let fixture = create_styles_fixture(
    Some(
      r#"{
          create as _create,
          createTheme as _createTheme,
          defineConsts as _defineConsts,
          defineVars as _defineVars,
          firstThatWorks as _firstThatWorks,
          keyframes as _keyframes,
          positionTry as _positionTry,
          props as _props
        }"#,
    ),
    None,
    Some(FixtureImportMap {
      create: "_create",
      create_theme: "_createTheme",
      define_consts: "_defineConsts",
      define_vars: "_defineVars",
      first_that_works: "_firstThatWorks",
      keyframes: "_keyframes",
      position_try: "_positionTry",
      props: "_props",
    }),
  );

  transform_and_snapshot(&fixture, "import_named_alias");
}

// --- importSources (string) ---
#[test]
fn import_sources_string() {
  let import_source = "foo-bar";
  let fixture = create_styles_fixture(Some("* as stylex"), Some(import_source), None);

  let output = transform_with_options(
    &fixture,
    Some(&mut get_default_opts()),
    Some(FileName::Real("/stylex/packages/vars.stylex.js".into())),
    Some(vec![ImportSources::Regular(import_source.to_string())]),
  );
  assert_snapshot!("import_sources_string", output);
}

// --- importSources (object) ---
#[test]
fn import_sources_object() {
  let fixture = create_styles_fixture(
    Some("{css, html}"),
    Some("react-strict-dom"),
    Some(FixtureImportMap {
      create: "css.create",
      create_theme: "css.createTheme",
      define_consts: "css.defineConsts",
      define_vars: "css.defineVars",
      first_that_works: "css.firstThatWorks",
      keyframes: "css.keyframes",
      position_try: "css.positionTry",
      props: "css.props",
    }),
  );

  let output = transform_with_options(
    &fixture,
    Some(&mut get_default_opts()),
    Some(FileName::Real("/stylex/packages/vars.stylex.js".into())),
    Some(vec![ImportSources::Named(NamedImportSource {
      r#as: "css".to_string(),
      from: "react-strict-dom".to_string(),
    })]),
  );
  assert_snapshot!("import_sources_object", output);
}

// --- [META-ONLY] import: default ---
#[test]
#[ignore = "META-ONLY test, not supported in OSS"]
fn meta_only_import_default() {
  let fixture = create_styles_fixture(Some("stylex"), Some("stylex"), None);
  transform_and_snapshot(&fixture, "meta_only_import_default");
}
