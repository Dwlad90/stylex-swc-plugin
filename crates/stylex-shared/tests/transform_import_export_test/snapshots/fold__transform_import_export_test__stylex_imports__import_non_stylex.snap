---
source: crates/stylex-shared/tests/transform_import_export_test/stylex_imports.rs
expression: output
---
import { foo, bar } from 'other';
